discovery/manager.go

```diff
@@ -78,12 +77,38 @@ var (
 		},
 		[]string{"name"},
 	)
+
+	registerMtx sync.Mutex
+	providers   = make(map[string]func(interface{}, log.Logger) Discoverer, 0)
 )
 
 func init() {
 	prometheus.MustRegister(failedConfigs, discoveredTargets, receivedUpdates, delayedUpdates, sentUpdates)
 }
 
+// RegisterDiscoverer registers a discoverer that can be used.
+func RegisterDiscoverer(name string, d func(interface{}, log.Logger) Discoverer) error {
+	registerMtx.Lock()
+	defer registerMtx.Unlock()
+	if _, ok := providers[name]; ok {
+		return fmt.Errorf("provider %s is already registered", name)
+	}
+	providers[name] = d
+	return nil
+}
+
+// GetDiscoverer return a discoverer that was registered with
+// RegisterDiscoverer.
+func GetDiscoverer(name string) (func(interface{}, log.Logger) Discoverer, error) {
+	registerMtx.Lock()
```

Comment:
I don't think init is ever done concurrently so we don't need any mutex here and above

Reply:
Do you think that all the init() functions are called in one go routing?

Reply:
yea I think so